scalameta/semantic/shared/src/main/scala/scala/meta/internal/semantic/codecs/package.scala

```diff
@@ -0,0 +1,102 @@
+            }
+          }
+          implicit val sugarEncoder = new ProtoEncoder[(m.Anchor, String), p.Sugar] {
+            override def toProto(e: (m.Anchor, String)): p.Sugar = e match {
+              case (m.Anchor(_, start, end), syntax) =>
+                p.Sugar(Some(p.Range(start, end)), syntax)
+            }
+          }
+          p.AttributedSource(path.toString,
+                             names.map(_.toProto[p.ResolvedName]).toSeq,
+                             messages.map(_.toProto[p.Message]),
+                             denotations.map(_.toProto[p.SymbolDenotation]).toSeq,
+                             sugars.map(_.toProto[p.Sugar]).toSeq)
+      }
+      override def fromProto(e: p.AttributedSource): m.AttributedSource = e match {
+        case p.AttributedSource(path, names, messages, denotations, sugars) =>
+          implicit val resolvedNameDecoder = new ProtoDecoder[(m.Anchor, m.Symbol), p.ResolvedName] {
+            override def fromProto(e: p.ResolvedName): (m.Anchor, m.Symbol) = e match {
+              case p.ResolvedName(Some(p.Range(start, end)), m.Symbol(symbol)) =>
+                m.Anchor(AbsolutePath(path), start, end) -> symbol
+              case _ =>
+                unreachable
+            }
+          }
+          implicit val messageDecoder = new ProtoDecoder[m.Message, p.Message] {
+            override def fromProto(e: p.Message): m.Message = e match {
+              case p.Message(Some(p.Range(start, end)), sev, message) =>
+                m.Message(m.Anchor(AbsolutePath(path), start, end), m.Severity.fromId(sev.value), message)
+              case _ =>
+                unreachable
+            }
+          }
+          implicit val symbolDenotationEncoder = new ProtoDecoder[(m.Symbol, m.Denotation), p.SymbolDenotation] {
+            override def fromProto(e: p.SymbolDenotation): (m.Symbol, m.Denotation) = e match {
+              case p.SymbolDenotation(m.Symbol(symbol), Some(p.Denotation(flags, name, info))) =>
+                symbol -> m.Denotation(flags, name, info)
+              case _ =>
+                unreachable
+            }
+          }
+          implicit val sugarDecoder = new ProtoDecoder[(m.Anchor, String), p.Sugar] {
+            override def fromProto(e: p.Sugar): (m.Anchor, String) = e match {
+              case p.Sugar(Some(p.Range(start, end)), syntax) =>
+                m.Anchor(AbsolutePath(path), start, end) -> syntax
+              case _ =>
+                unreachable
+            }
+          }
+          m.AttributedSource(AbsolutePath(path),
+                             names.map(_.toMeta[(m.Anchor, m.Symbol)]).toMap,
+                             messages.map(_.toMeta[m.Message]).toList,
+                             denotations.map(_.toMeta[(m.Symbol, m.Denotation)]).toMap,
+                             sugars.map(_.toMeta[(m.Anchor, String)]).toMap)
+      }
+    }
+
+  implicit val DatabaseCodec: ProtoCodec[m.Database, p.Database] =
```

Comment:
Alternatively, we could wrap all of the encoders/decoders into a class

```scala
class DatabaseCodecs(sourceRoot: AbsolutePath) {
  implicit val DatabaseCodec: ProtoCodec[m.Database, p.Database]
}
```

This would avoid having to pass around the sourceRoot between codecs.